upgrade/upgrades: TestMigrationWithFailures failed

upgrade/upgrades.TestMigrationWithFailures [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9055212?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9055212?buildTab=artifacts#/) on master @ [0c6ccaf5c6fc8f94bb2a1e6132e44fe9e82bd9ed](https://github.com/cockroachdb/cockroach/commits/0c6ccaf5c6fc8f94bb2a1e6132e44fe9e82bd9ed):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3c7d7d7]

```
Stack: 

```
goroutine 9644452 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x3?, {0x5bee6c8, 0xc0099ecf60})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x43065a0, 0x81feb50})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/server.(*refreshInstanceSessionListener).OnSessionDeleted.func1({0xc0032ee400, 0x0})
	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:472 +0x117
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestMigrationWithFailures_adding_columns1172456435
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 847787294144790529
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:517: -- test log scope end --
=== RUN   TestMigrationWithFailures/adding_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestMigrationWithFailures_adding_index2454347597
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 847787325077946369
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:517: -- test log scope end --
=== RUN   TestMigrationWithFailures/fail_adding_columns
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestMigrationWithFailures_fail_adding_columns3385476765
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 847787350060138497
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:468: waiting for the migration job to observe a mutation
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:482: waiting for the schema job to reach the cancel status
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:517: -- test log scope end --
=== RUN   TestMigrationWithFailures/fail_adding_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestMigrationWithFailures_fail_adding_index511149566
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 847787385062424577
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:468: waiting for the migration job to observe a mutation
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:482: waiting for the schema job to reach the cancel status
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationWithFailures.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25335

upgrade/upgrades.TestMigrationWithFailures [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9139159?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9139159?buildTab=artifacts#/) on master @ [9aa4f3485c75397e9ce0edabbf22c6343e1a901a](https://github.com/cockroachdb/cockroach/commits/9aa4f3485c75397e9ce0edabbf22c6343e1a901a):


```
=== RUN   TestMigrationWithFailures
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures1597166461
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestMigrationWithFailures
    schema_changes_external_test.go:519: -- test log scope end --
I230319 21:05:45.439778 48592 3@pebble/event.go:706  [n?,s?,pebble] 309  [JOB 1] MANIFEST created 000001
I230319 21:05:45.439978 48592 3@pebble/event.go:734  [n?,s?,pebble] 310  [JOB 1] WAL created 000002
I230319 21:05:45.440027 48592 3@pebble/event.go:702  [n?,s?,pebble] 311  upgraded to format version: 002
I230319 21:05:45.440054 48592 3@pebble/event.go:702  [n?,s?,pebble] 312  upgraded to format version: 003
I230319 21:05:45.440071 48592 3@pebble/event.go:702  [n?,s?,pebble] 313  upgraded to format version: 004
I230319 21:05:45.440086 48592 3@pebble/event.go:702  [n?,s?,pebble] 314  upgraded to format version: 005
I230319 21:05:45.440106 48592 3@pebble/event.go:702  [n?,s?,pebble] 315  upgraded to format version: 006
I230319 21:05:45.440122 48592 3@pebble/event.go:702  [n?,s?,pebble] 316  upgraded to format version: 007
I230319 21:05:45.440138 48592 3@pebble/event.go:702  [n?,s?,pebble] 317  upgraded to format version: 008
--- FAIL: TestMigrationWithFailures (106.57s)
=== RUN   TestMigrationWithFailures/fail_adding_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures_fail_adding_index563860154
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 849312123107049473
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:468: waiting for the migration job to observe a mutation
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:482: waiting for the schema job to reach the cancel status
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:515: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/schema_changes_external_test.go:515
        	Error:      	"[{INFO %!s(int64=1679259930519045000) %!s(int64=35499) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-index operation as the schema change already exists. n1,job=MIGRATION id=849312130222522369,upgrade=1000022.2-90 %!s(uint64=745) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679259930507321000) %!s(int64=35499) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849312130222522369,upgrade=1000022.2-90 %!s(uint64=740) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679259894055445000) %!s(int64=25785) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849312130222522369,upgrade=1000022.2-90 %!s(uint64=440) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1}]" should have 1 item(s), but has 3
        	Test:       	TestMigrationWithFailures/fail_adding_index
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures_fail_adding_index563860154
    --- FAIL: TestMigrationWithFailures/fail_adding_index (41.76s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationWithFailures.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestMigrationWithFailures [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9170520?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9170520?buildTab=artifacts#/) on master @ [1e9899da8cb250a4e560b280beb2b0805ee75a78](https://github.com/cockroachdb/cockroach/commits/1e9899da8cb250a4e560b280beb2b0805ee75a78):


```
=== RUN   TestMigrationWithFailures
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures495018895
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestMigrationWithFailures
    schema_changes_external_test.go:519: -- test log scope end --
--- FAIL: TestMigrationWithFailures (68.19s)
=== RUN   TestMigrationWithFailures/skip_adding_columns
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures_skip_adding_columns3150689080
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 849779970167504897
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:468: waiting for the migration job to observe a mutation
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:482: waiting for the schema job to reach the cancel status
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:515: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/schema_changes_external_test.go:515
        	Error:      	"[{INFO %!s(int64=1679402706515917000) %!s(int64=34066) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-index operation as the schema change already exists. n1,job=MIGRATION id=849779977681043457,upgrade=1000022.2-90 %!s(uint64=519) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679402706514263000) %!s(int64=34066) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849779977681043457,upgrade=1000022.2-90 %!s(uint64=515) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679402669783742000) %!s(int64=23839) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849779977681043457,upgrade=1000022.2-90 %!s(uint64=225) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1}]" should have 1 item(s), but has 3
        	Test:       	TestMigrationWithFailures/skip_adding_columns
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/99d7f9d8b90eb33a7ba0b803c5a3f9d7/logTestMigrationWithFailures_skip_adding_columns3150689080
    --- FAIL: TestMigrationWithFailures/skip_adding_columns (42.41s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationWithFailures.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestMigrationWithFailures [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9171084?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9171084?buildTab=artifacts#/) on master @ [5f5cf265a8717da8ee327f41c0d06836d11ac8ec](https://github.com/cockroachdb/cockroach/commits/5f5cf265a8717da8ee327f41c0d06836d11ac8ec):


```
=== RUN   TestMigrationWithFailures
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/effffcb4dff575c191a0d0aec2225271/logTestMigrationWithFailures3764211966
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestMigrationWithFailures
    schema_changes_external_test.go:519: -- test log scope end --
--- FAIL: TestMigrationWithFailures (117.80s)
=== RUN   TestMigrationWithFailures/fail_adding_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/effffcb4dff575c191a0d0aec2225271/logTestMigrationWithFailures_fail_adding_index4136180422
    test_log_scope.go:79: use -show-logs to present logs inline
    schema_changes_external_test.go:397: trying migration, expecting to fail
    schema_changes_external_test.go:406: intercepting the schema job
    schema_changes_external_test.go:422: intercepted schema change job: 849784438042132481
    schema_changes_external_test.go:429: canceling the migration job
    schema_changes_external_test.go:434: waiting for the migration job to finish.
    schema_changes_external_test.go:463: retrying migration, expecting to succeed
    schema_changes_external_test.go:468: waiting for the migration job to observe a mutation
    schema_changes_external_test.go:472: resuming the schema change job
    schema_changes_external_test.go:482: waiting for the schema job to reach the cancel status
    schema_changes_external_test.go:497: waiting for the new migration job to complete.
    schema_changes_external_test.go:515: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/schema_changes_external_test.go:515
        	Error:      	"[{INFO %!s(int64=1679404069481397000) %!s(int64=28720) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-index operation as the schema change already exists. n1,job=MIGRATION id=849784445261479937,upgrade=1000022.2-90 %!s(uint64=518) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679404069479594000) %!s(int64=28720) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849784445261479937,upgrade=1000022.2-90 %!s(uint64=516) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1} {INFO %!s(int64=1679404033059504000) %!s(int64=18245) upgrade/upgrades/schema_changes.go %!s(int64=119) skipping jobs-add-columns operation as the schema change already exists. n1,job=MIGRATION id=849784445261479937,upgrade=1000022.2-90 %!s(uint64=224) %!s(bool=false) DEV %!s(uint32=0) %!s(uint32=0) %!s(uint32=0) 1}]" should have 1 item(s), but has 3
        	Test:       	TestMigrationWithFailures/fail_adding_index
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/effffcb4dff575c191a0d0aec2225271/logTestMigrationWithFailures_fail_adding_index4136180422
    --- FAIL: TestMigrationWithFailures/fail_adding_index (42.95s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationWithFailures.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/schema_changes_external_test.go
+++ b/pkg/upgrade/upgrades/schema_changes_external_test.go
@@ -69,7 +69,6 @@ type schemaChangeTestCase struct {
 // exponential backoff to the system.jobs table, but was retrofitted to prevent
 // regressions.
 func TestMigrationWithFailures(t *testing.T) {
-	skip.WithIssue(t, 98564)
 	const createTableBefore = `
 CREATE TABLE test.test_table (
 	id                INT8      DEFAULT unique_rowid() PRIMARY KEY,
@@ -174,7 +173,6 @@ CREATE TABLE test.test_table (
 // alters a column in a table multiple times with failures at different stages
 // of the migration.
 func TestMigrationWithFailuresMultipleAltersOnSameColumn(t *testing.T) {
-	skip.WithIssue(t, 98991)
 
 	const createTableBefore = `
 CREATE TABLE test.test_table (
@@ -245,9 +243,12 @@ func testMigrationWithFailures(
 
 	skip.UnderRace(t, "very slow")
 
-	// We're going to be migrating from the minimum supported version to the current version.
-	startCV := clusterversion.ByKey(clusterversion.BinaryMinSupportedVersionKey)
-	endCV := clusterversion.ByKey(clusterversion.BinaryVersionKey)
+	// We're going to be migrating from the minimum supported version to the
+	// "next" version. We'll be injecting the migration for the next version.
+	startKey := clusterversion.BinaryMinSupportedVersionKey
+	startCV := clusterversion.ByKey(startKey)
+	endCV := startCV
+	endCV.Internal += 2
 
 	// The tests follows the following procedure.
 	//
@@ -340,6 +341,7 @@ func testMigrationWithFailures(
 						Server: &server.TestingKnobs{
 							DisableAutomaticVersionUpgrade: make(chan struct{}),
 							BinaryVersionOverride:          startCV,
+							BootstrapVersionKeyOverride:    startKey,
 						},
 						JobsTestingKnobs: jobsKnobs,
 						SQLExecutor: &sql.ExecutorTestingKnobs{

